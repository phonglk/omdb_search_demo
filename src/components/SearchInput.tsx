import React, {
  ChangeEventHandler,
  KeyboardEventHandler,
  useCallback,
} from 'react';
import styled from 'styled-components';

const Input = styled.input`
  line-height: 20px;
  flex: 1 1 auto;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  font-size: 16px;
  font-weight: 200;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.8);
  filter: drop-shadow(0 3px 3px rgba(0, 0, 0, 0.1));
`;

const Wrapper = styled.div`
  display: flex;
  flex: 0 0 50px;
`;

const Button = styled.button`
  border-radius: 5px;
  flex: 0 0 auto;
  margin-left: 10px;
  height: 100%;
  display: block;
  background: #090979;
  border: none;
  color: white;
  padding: 5px 20px;
  font-size: 16px;
  cursor: pointer;
  transition: all 0.3s;
  &:hover {
    background-color: #00d4ff;
  }
  &:disabled {
    background-color: rgba(0, 0, 0, 0.5);
    cursor: not-allowed;
  }
`;

type Props = {
  handleSearch: () => void;
  searchTerm: string;
  handleInputUpdate: (term: string) => void;
  disabled: boolean;
};

export default function SearchInput({
  searchTerm,
  handleSearch,
  handleInputUpdate,
  disabled,
}: Props) {
  const handleKeyDown: KeyboardEventHandler = useCallback(
    (event) => {
      if (event.key === 'Enter') handleSearch();
    },
    [handleSearch]
  );

  const handleChange: ChangeEventHandler<HTMLInputElement> = useCallback(
    (event) => {
      handleInputUpdate(event.target.value);
    },
    [handleInputUpdate]
  );

  return (
    <Wrapper>
      <Input
        type="text"
        value={searchTerm}
        onKeyDown={handleKeyDown}
        onChange={handleChange}
        disabled={disabled}
      />
      <Button onClick={handleSearch} disabled={disabled}>
        Search
      </Button>
    </Wrapper>
  );
}
