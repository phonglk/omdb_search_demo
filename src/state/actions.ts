import { useMemo } from 'react';
import { Movie } from '../types';
import { useStoreContext } from './store';

export enum ActionType {
  UPDATE_SEARCH_TERM,
  SEARCH_MOVIE,
  SEARCH_DONE,
}

export type Action =
  | { type: ActionType.UPDATE_SEARCH_TERM; payload: string }
  | { type: ActionType.SEARCH_MOVIE }
  | { type: ActionType.SEARCH_DONE; payload: Movie[] };

export const useActions = () => {
  const { dispatch } = useStoreContext();

  return useMemo(
    () => ({
      updateSearchTerm: (term: string) =>
        dispatch({
          type: ActionType.UPDATE_SEARCH_TERM,
          payload: term,
        }),
      searchMovie: () => dispatch({ type: ActionType.SEARCH_MOVIE }),
      searchDone: (result: Movie[]) =>
        dispatch({ type: ActionType.SEARCH_DONE, payload: result }),
    }),
    [dispatch]
  );
};
