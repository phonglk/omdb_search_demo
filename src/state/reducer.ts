import { State } from '../types';
import { Action, ActionType } from './actions';

export const initialState: State = {
  searchTerm: '',
  isSearching: false,
  searchResult: [],
};

export default function reducer(state: State, action: Action): State {
  switch (action.type) {
  case ActionType.UPDATE_SEARCH_TERM:
    return { ...state, searchTerm: action.payload };
  case ActionType.SEARCH_MOVIE:
    return { ...state, isSearching: true };
  case ActionType.SEARCH_DONE:
    return { ...state, isSearching: false, searchResult: action.payload };
  default:
    return state;
  }
}
