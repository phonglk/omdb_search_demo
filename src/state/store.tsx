import React from 'react';
import _ from 'lodash';
import { createContext, Dispatch, useContext, useReducer } from 'react';
import { ComponentProps, State } from '../types';
import { Action } from './actions';
import reducer, { initialState } from './reducer';

type StoreContextValue = {
  state: State;
  dispatch: Dispatch<Action>;
};

export const StoreContext = createContext<StoreContextValue>({
  state: initialState,
  dispatch: _.identity,
});

export const useStoreReducer = () => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return { state, dispatch };
};

export const StoreProvider = (props: ComponentProps) => {
  const { state, dispatch } = useStoreReducer();
  return (
    <StoreContext.Provider value={{ state, dispatch }}>
      {props.children}
    </StoreContext.Provider>
  );
};

export const useStoreContext = () => useContext(StoreContext);

export const useStoreState = () => {
  const { state } = useStoreContext();
  return state;
};
