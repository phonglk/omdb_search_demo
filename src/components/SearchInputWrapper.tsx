import React from 'react';
import { useActions } from '../state/actions';
import { useStoreState } from '../state/store';
import search from '../utils.ts/search';
import SearchInput from './SearchInput';

export default function SearchInputWrapper() {
  const { updateSearchTerm, searchMovie, searchDone } = useActions();
  const { searchTerm, isSearching } = useStoreState();

  const handleSearch = async () => {
    searchMovie();
    try {
      const movies = await search(searchTerm);
      console.log(movies);

      searchDone(movies);
    } catch (error: any) {
      alert(error.message);
      searchDone([]);
    }
  };

  return (
    <SearchInput
      handleSearch={handleSearch}
      handleInputUpdate={updateSearchTerm}
      searchTerm={searchTerm}
      disabled={isSearching}
    />
  );
}
