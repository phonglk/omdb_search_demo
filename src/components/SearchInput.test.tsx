import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import SearchInputWrapper from './SearchInputWrapper';
import { StoreProvider } from '../state/store';
import SearchInput from './SearchInput';
import _ from 'lodash';

beforeEach(() => {
  jest.spyOn(global, 'fetch');
});

afterEach(() => {
  jest.restoreAllMocks();
});

const setup = () => {
  render(
    <StoreProvider>
      <SearchInputWrapper />
    </StoreProvider>
  );
  const input =
    screen.getByPlaceholderText<HTMLInputElement>('Search title...');
  fireEvent.change(input, { target: { value: 'spider' } });

  const searchBtn = screen.getByText('Search');

  return { input, searchBtn };
};

test('Input update text', () => {
  const { input } = setup();
  expect(input.value).toBe('spider');
});

test('Call API when user click search', () => {
  const { searchBtn } = setup();
  fireEvent.click(searchBtn);
  expect(global.fetch).toBeCalledWith(
    'https://omdbapi.com/?s=spider&apikey=TEST'
  );
});

test('Call API when user click search', () => {
  const { input } = setup();
  fireEvent.keyDown(input, { key: 'Enter' });
  expect(global.fetch).toBeCalledWith(
    'https://omdbapi.com/?s=spider&apikey=TEST'
  );
});

test('SearchInput snapshot test', () => {
  const { container } = render(
    <SearchInput
      handleSearch={_.identity}
      searchTerm={'search'}
      handleInputUpdate={_.identity}
      disabled={false}
    />
  );
  expect(container).toMatchSnapshot();
});
