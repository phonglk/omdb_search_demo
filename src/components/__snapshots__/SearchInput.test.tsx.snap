// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput snapshot test 1`] = `
<div>
  <div
    class="sc-gsnTZi fvPgcZ"
  >
    <input
      class="sc-bczRLJ dSlICL"
      placeholder="Search title..."
      type="text"
      value="search"
    />
    <button
      class="sc-dkzDqf gauXII"
    >
      Search
    </button>
  </div>
</div>
`;
